pkg/proto/proto.go

```diff
@@ -64,6 +89,138 @@ func ProtocolVersion() Version {
 	return NewVersion(major, minor, patch)
 }
 
+// ParseMessage parses a message from the given data. The 'f' function parameter is used to parse the message payload.
+func ParseMessage(data []byte, contentID PeerMessageID, name string, f func(payload []byte) error) error {
+	l, err := safecast.ToUint32(len(data))
+	if err != nil {
+		return fmt.Errorf("%s: %w", name, err)
+	}
+	if l < headerSizeWithoutPayload {
+		return fmt.Errorf("%s: invalid data size %d, expected at least %d",
+			name, len(data), headerSizeWithoutPayload)
+	}
+	var h Header
+	if ubErr := h.UnmarshalBinary(data); ubErr != nil {
+		return fmt.Errorf("%s: %w", name, ubErr)
+	}
+	if vErr := h.Validate(contentID); vErr != nil {
+		return fmt.Errorf("%s: %w", name, vErr)
+	}
+	if exp, act := int(h.Length+msgLenSize), len(data); act < exp { // Add the length of the MSG_LEN field itself.
+		return fmt.Errorf("%s: expected data at least %d, found %d", name, exp, act)
+	}
+	if h.payloadLength > 0 {
+		if fErr := f(data[headerSizeWithPayload : headerSizeWithPayload+h.payloadLength]); fErr != nil {
+			return fmt.Errorf("%s: payload error: %w", name, fErr)
+		}
+	}
+	return nil
+}
+
+type ChecksumReader struct {
+	r io.Reader
+	h hash.Hash
+}
+
+func NewChecksumReader(r io.Reader) (*ChecksumReader, error) {
+	h, err := blake2b.New256(nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create ChecksumReader: %w", err)
+	}
+	return &ChecksumReader{
+		r: r,
+		h: h,
+	}, nil
+}
+
+// Read reads data from the underlying reader and updates the checksum.
+func (r *ChecksumReader) Read(p []byte) (int, error) {
+	n, err := r.r.Read(p)
+	if n > 0 {
+		if wn, wErr := r.h.Write(p[:n]); wErr != nil {
+			return wn, fmt.Errorf("failed to update hasher: %w", wErr)
+		}
+	}
+	return n, err
```

Comment:
Is it analogous to `io.TeeReader` as I can see

Reply:
I was thinking about implementing a MessageReader also, it will incapsulate whole messages reading. In this case the TeeReader will be useful as a part of complex reading process, I think.